### PR TITLE
Add some pipeline invariants to the signature and create a cli subcommand to sign a pipeline

### PR DESCRIPTION
**Pull request by triarius**
The pipeline invariants are:
- pipeline slug
- organization slug
- repository

An agent configured to verify signatures will now check that the pipeline slug, organization slug and repository in the accepted job definition are the same as the ones in the signature.

The cli subcommand
```
buildkite-agnet tool sign --jwks-file-path <keyfile> --signing-key-id <key-id> --pipeline-slug <pipeline-slug> --organization-slug <organization-slug> --repo <repository> <pipeline-yaml-file>
```
can be used to sign the initial pipeline. Its output may be pasted into the Steps Editor in the Buildkite Web UI so that the initial pipeline may be verified as well.

## Implementation details
I created a new struct `CommandStepWithInvariants` and moved the `SignedFielder` implementation from `CommandStep` to it. I went back and forth on a few ways to do this, but I think it's easier to reason about if the marshalling of fields to sign happens in one place.

### CODE REVIEW

File: clicommand/commands.go [hunk 1]
```diff
@@ -82,6 +82,7 @@ var BuildkiteAgentCommands = []cli.Command{
 		Usage: "Utility commands, intended for users and operators of the agent to run directly on their machines, and not as part of a Buildkite job",
 		Subcommands: []cli.Command{
 			KeygenCommand,
+			ToolSignCommand,
```
Comment (moskyb) [inline clicommand/commands.go:85]: should we either rename `KeyGen` to `ToolKeygen` or rename `ToolSign` to `Sign`?

File: clicommand/pipeline_upload.go [hunk 2]
```diff
@@ -116,7 +116,7 @@ var PipelineUploadCommand = cli.Command{
 		},
 		cli.BoolFlag{
 			Name:   "no-interpolation",
-			Usage:  "Skip variable interpolation the pipeline when uploaded",
+			Usage:  "Skip variable interpolation into the pipeline prior to upload",
```
Comment (moskyb) [inline clicommand/pipeline_upload.go:119]: 👍

File: internal/pipeline/steps.go [hunk 3]
```diff
@@ -137,18 +137,23 @@ func stepByKeyInference(o *ordered.MapSA) (Step, error) {
 // sign adds signatures to each command step (and recursively to any command
 // steps that are within group steps. The steps are mutated directly, so an
 // error part-way through may leave some steps un-signed.
-func (s Steps) sign(env map[string]string, key jwk.Key) error {
+func (s Steps) sign(key jwk.Key, env map[string]string, pInv *PipelineInvariants) error {
```
Comment (moskyb) [inline internal/pipeline/steps.go:140]: 👍

**Review by moskyb — CHANGES_REQUESTED** [review]
looks good! there's one that i'm not sure about on re: using `environ`, but other than that this looks geat

File: clicommand/pipeline_upload.go [hunk 4]
```diff
@@ -277,12 +277,18 @@ var PipelineUploadCommand = cli.Command{
 		if cfg.JWKSFilePath != "" {
 			l.Warn("Pipeline signing is experimental and the user interface might change! Also it might not work, it might sign the pipeline only partially, or it might eat your pet dog. You have been warned!")
 
-			key, err := loadSigningKey(cfg)
+			pInv := &pipeline.PipelineInvariants{
+				OrganizationSlug: environ.GetStringDefaultEmpty("BUILDKITE_ORGANIZATION_SLUG"),
+				PipelineSlug:     environ.GetStringDefaultEmpty("BUILDKITE_PIPELINE_SLUG"),
+				Repository:       environ.GetStringDefaultEmpty("BUILDKITE_REPO"),
```
[the remark this answers is not in the document]
Comment (triarius) [reply clicommand/pipeline_upload.go:283]: Nope, it's me that's missing the context. Good catch!

File: clicommand/commands.go (hunk 1 above)
Comment (triarius) [reply clicommand/commands.go:85]: I think you know which one I will choose 😂

File: internal/pipeline/steps.go (hunk 3 above)
Comment (triarius) [reply internal/pipeline/steps.go:140]: I feel like there's a another place that's like this. I'll see if I can put the key consistently as the first arg everywhere.

**Review by moskyb — APPROVED** [review]

**Pull request merged**